experimental/devices/hd44780/lcd_hd44780_4bit.go

```diff
@@ -0,0 +1,230 @@
+// Copyright 2018 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+// Package hd44780 controls the Hitachi LCD display chipset HD-44780
+//
+// Datasheet
+//
+// https://www.sparkfun.com/datasheets/LCD/HD44780.pdf
+package hd44780
+
+import (
+	"time"
+
+	"periph.io/x/periph/conn/gpio"
+)
+
+// LineTwo offset for the second line in the LCD buffer.
+const LineTwo = 0x40
+
+// LCD4 is the 4-bit addressing device for HD-44780
+type LCD4 struct {
+	// data pins
+	dataPins []gpio.PinOut
+
+	// register select pin
+	rsPin gpio.PinOut
+
+	// enable pin
+	enablePin gpio.PinOut
+}
+
+// NewLCD4Bit creates and initializes the LCD device
+//	data - references to data pins
+//	rs - rs pin
+//	e - strobe pin
+func NewLCD4Bit(data []gpio.PinOut, rs, e gpio.PinOut) (*LCD4, error) {
+	dev := LCD4{
+		dataPins:  data,
+		enablePin: e,
+		rsPin:     rs,
+	}
+	if err := dev.Init(); err != nil {
+		return nil, err
+	}
+	return &dev, nil
+}
+
+var resetSequence = [][]uint{
+	{0x03, 50}, // init 1-st cycle
+	{0x03, 10}, // init 2-nd cycle
+	{0x03, 10}, // init 3-rd cycle
+	{0x02, 10}, // init finish
+}
+
+var initSequence = [][]uint{
+	{0x14, 0},    // 4-bit mode, 2 lines, 5x7 chars high
+	{0x10, 0},    // disable display
+	{0x01, 2000}, // clear screen
+	{0x06, 0},    // cursor shift right, no display move
+	{0x0c, 0},    // enable display no cursor
+	{0x01, 2000}, // clear screen
+	{0x02, 2000}, // cursor home
+}
+
+// Init initializes the HC-44780 chipset, clears the screen buffer and moves cursor to the
+// home of screen (line 0, column 0)
+func (r *LCD4) Init() error {
+	if err := clearBits(r); err != nil {
+		return err
+	}
+
+	delayMs(15)
+
+	if err := r.rsPin.Out(gpio.Low); err != nil {
+		return err
+	}
+	if err := r.enablePin.Out(gpio.Low); err != nil {
+		return err
+	}
+
+	wrapper := func(seq [][]uint, f func(_lcd *LCD4, _data uint8) error) error {
+		for _, v := range seq {
+			if err := f(r, uint8(v[0])); err != nil {
+				return err
+			}
+			if v[1] > 0 {
+				delayUs(v[1])
+			}
+		}
+		return nil
+	}
+
+	if err := wrapper(resetSequence, write4Bits); err != nil {
+		return err
+	}
+
+	if err := wrapper(initSequence, writeInstruction); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Cls clears the LCD screen
+func (r *LCD4) Cls() error {
+	if err := writeInstruction(r, 0x01); err != nil {
+		return err
+	}
+	delayMs(2)
```

Comment:
Is the delay needed? Instead you could note the last timestamp of the command inside Dev and then wait for a delay on every function call, e.g.
```
type Dev struct {
 ...
  waitUntil time.Time
}

func (d *Dev) Clear() error {
  if s := time.Until(d.waitUntil); s > 0 {
    time.Sleep(s)
  }
  ...
  d.waitUntil = time.Now().Add(time.Millisecond*2)
  return nil
}
```

and in all the public methods, do

```
  if s := time.Until(d.waitUntil); s > 0 {
    time.Sleep(s)
  }
```